src/utils/periodic_utils.cpp

```diff
@@ -0,0 +1,65 @@
+/*
+ * MRCPP, a numerical library based on multiresolution analysis and
+ * the multiwavelet basis which provide low-scaling algorithms as well as
+ * rigorous error control in numerical computations.
+ * Copyright (C) 2020 Stig Rune Jensen, Jonas Juselius, Luca Frediani and contributors.
+ *
+ * This file is part of MRCPP.
+ *
+ * MRCPP is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MRCPP is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with MRCPP.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * For information on the complete list of contributors to MRCPP, see:
+ * <https://mrcpp.readthedocs.io/>
+ */
+
+#include "periodic_utils.h"
+#include "trees/NodeIndex.h"
+
+#include <cmath>
+
+namespace mrcpp {
+namespace periodic {
+
+template <int D> void indx_manipulation(NodeIndex<D> &idx, const std::array<bool, D> &periodic) {
+    int translation[D];
+    int two_n = 1 << idx.getScale();
+    for (auto i = 0; i < D; i++) {
+        if (periodic[i]) {
+            translation[i] = idx.getTranslation(i);
+            if (translation[i] >= two_n) translation[i] = translation[i] % two_n;
+            if (translation[i] < 0) translation[i] = (translation[i] + 1) % two_n + two_n - 1;
+        }
+    }
+    idx.setTranslation(translation);
+}
```

Comment:
`translation[i]` remains uninitialized if `periodic[i] == false`

Reply:
Also, the bit shift will not work if `scale` is negative. This should not happen in periodic calculations, but I think you can still add a guard for this.

Reply:
Fixed :)